File: packages/go/behavioral.go
package gof

import (
	"errors"
	"fmt"
	"iter"
	"math"
	"sync"
)

// ErrNoHandler is returned by a [Chain] whose handlers all declined and that was
// given no fallback.
var ErrNoHandler = errors.New("gof: no handler answered and no fallback was given")

// Chain is Chain of Responsibility: each handler either answers or calls next.
//
// Pass nil for fallback to get [ErrNoHandler] when nothing answers; the returned
// func reports it through the error result.
//
//	route := gof.Chain([]func(int, func() string) string{
//		func(level int, next func() string) string { if level == 1 { return "bot" }; return next() },
//	}, func(int) string { return "queue" })
func Chain[Req, Res any](
	handlers []func(Req, func() Res) Res,
	fallback func(Req) Res,
) func(Req) (Res, error) {
	return func(request Req) (result Res, err error) {
		var step func(int) Res
		step = func(i int) Res {
			if i >= len(handlers) {
				if fallback == nil {
					err = ErrNoHandler
					var zero Res
					return zero
				}
				return fallback(request)
			}
			return handlers[i](request, func() Res { return step(i + 1) })
		}
		return step(0), err
	}
}

// Command is one undoable operation, built from two funcs so the simple case needs
// no type of its own. A Command with a nil Undo still runs, but is not tracked.
type Command[T any] struct {
	Execute func() T
	Undo    func()
}

// CommandBus runs commands and keeps the undo and redo history.
//
// Use the package-level [Run] rather than a method: a Go method cannot introduce a
// type parameter, so the command's result type has to come from a function.
type CommandBus struct {
	done   []Command[any]
	undone []Command[any]
}

// NewCommandBus returns an empty bus.
func NewCommandBus() *CommandBus { return &CommandBus{} }

// Run executes cmd, records it if it can be undone, and returns its result.
func Run[T any](bus *CommandBus, cmd Command[T]) T {
	result := cmd.Execute()
	bus.undone = bus.undone[:0]
	if cmd.Undo != nil {
		bus.done = append(bus.done, Command[any]{
			Execute: func() any { return cmd.Execute() },
			Undo:    cmd.Undo,
		})
	}
	return result
}

// Undo reverses the last undoable command, reporting whether there was one.
func (b *CommandBus) Undo() bool {
	if len(b.done) == 0 {
		return false
	}
	cmd := b.done[len(b.done)-1]
	b.done = b.done[:len(b.done)-1]
	cmd.Undo()
	b.undone = append(b.undone, cmd)
	return true
}

// Redo re-runs the last undone command, reporting whether there was one.
func (b *CommandBus) Redo() bool {
	if len(b.undone) == 0 {
		return false
	}
	cmd := b.undone[len(b.undone)-1]
	b.undone = b.undone[:len(b.undone)-1]
	cmd.Execute()
	b.done = append(b.done, cmd)
	return true
}

// CanUndo reports whether there is anything to undo.
func (b *CommandBus) CanUndo() bool { return len(b.done) > 0 }

// CanRedo reports whether there is anything to redo.
func (b *CommandBus) CanRedo() bool { return len(b.undone) > 0 }

// Subject is Observer: one typed channel of values. Subscribe returns the
// unsubscribe func.
//
// A Go channel plus a goroutine is the idiomatic fan-out and is better when the
// consumers are concurrent. This is for synchronous callbacks, where a channel
// would mean a goroutine per listener and no ordering guarantee.
type Subject[T any] struct {
	mu        sync.Mutex
	listeners []*func(T)
}

// Subscribe registers listener and returns a func that removes it. Calling that
// func twice is harmless.
func (s *Subject[T]) Subscribe(listener func(T)) func() {
	s.mu.Lock()
	defer s.mu.Unlock()
	handle := &listener
	s.listeners = append(s.listeners, handle)
	return func() {
		s.mu.Lock()
		defer s.mu.Unlock()
		for i, existing := range s.listeners {
			if existing == handle {
				s.listeners = append(s.listeners[:i], s.listeners[i+1:]...)
				return
			}
		}
	}
}

// Emit calls every listener with value, in subscription order.
func (s *Subject[T]) Emit(value T) {
	s.mu.Lock()
	// Copy first: a listener may unsubscribe while being called.
	current := make([]*func(T), len(s.listeners))
	copy(current, s.listeners)
	s.mu.Unlock()
	for _, listener := range current {
		(*listener)(value)
	}
}

// Len reports how many listeners are subscribed.
func (s *Subject[T]) Len() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.listeners)
}

// Channel is a named, typed event for a [Mediator]. Declare it once and share it;
// the name is only for debugging.
type Channel[T any] struct{ Name string }

// NewChannel returns a Channel carrying values of type T.
func NewChannel[T any](name string) *Channel[T] { return &Channel[T]{Name: name} }

// Mediator is a hub components talk to instead of to each other. Channels carry
// their payload type, so a listener cannot be attached to the wrong event.
//
// [On] and [Emit] are package-level funcs, not methods, because a Go method cannot
// introduce a type parameter.
type Mediator struct {
	mu       sync.Mutex
	subjects map[any]any
}

// NewMediator returns an empty hub.
func NewMediator() *Mediator { return &Mediator{subjects: make(map[any]any)} }

func subjectFor[T any](m *Mediator, ch *Channel[T]) *Subject[T] {
	m.mu.Lock()
	defer m.mu.Unlock()
	if existing, ok := m.subjects[ch]; ok {
		return existing.(*Subject[T])
	}
	created := &Subject[T]{}
	m.subjects[ch] = created
	return created
}

// On subscribes listener to ch and returns the unsubscribe func.
func On[T any](m *Mediator, ch *Channel[T], listener func(T)) func() {
	return subjectFor(m, ch).Subscribe(listener)
}

// Emit sends payload to every listener on ch.
func Emit[T any](m *Mediator, ch *Channel[T], payload T) {
	subjectFor(m, ch).Emit(payload)
}

// History is Memento: undo and redo over snapshots of state.
//
// Snapshots are stored as given. Pass [WithSnapshot] when the state is mutated in
// place, or the history will fill with the same mutated value.
type History[T any] struct {
	limit    int
	snapshot func(T) T
	past     []T
	future   []T
	current  T
}

// HistoryOption configures a [History].
type HistoryOption[T any] func(*History[T])

// WithLimit caps how many past states are kept; older ones are dropped.
func WithLimit[T any](limit int) HistoryOption[T] {
	return func(h *History[T]) { h.limit = limit }
}

// WithSnapshot sets the copy func applied to every saved state.
func WithSnapshot[T any](snapshot func(T) T) HistoryOption[T] {
	return func(h *History[T]) { h.snapshot = snapshot }
}

// NewHistory returns a History starting at initial.
func NewHistory[T any](initial T, options ...HistoryOption[T]) *History[T] {
	h := &History[T]{limit: math.MaxInt, snapshot: func(value T) T { return value }}
	for _, option := range options {
		option(h)
	}
	h.current = h.snapshot(initial)
	return h
}

// Current returns the state now.
func (h *History[T]) Current() T { return h.current }

// Save records the current state and moves to the new one, dropping any redo future.
func (h *History[T]) Save(state T) {
	h.past = append(h.past, h.current)
	if len(h.past) > h.limit {
		h.past = h.past[1:]
	}
	h.future = h.future[:0]
	h.current = h.snapshot(state)
}

// Undo steps back, reporting whether there was anywhere to go.
func (h *History[T]) Undo() bool {
	if len(h.past) == 0 {
		return false
	}
	h.future = append(h.future, h.current)
	h.current = h.past[len(h.past)-1]
	h.past = h.past[:len(h.past)-1]
	return true
}

// Redo steps forward, reporting whether there was anywhere to go.
func (h *History[T]) Redo() bool {
	if len(h.future) == 0 {
		return false
	}
	h.past = append(h.past, h.current)
	h.current = h.future[len(h.future)-1]
	h.future = h.future[:len(h.future)-1]
	return true
}

// CanUndo reports whether Undo would move.
func (h *History[T]) CanUndo() bool { return len(h.past) > 0 }

// CanRedo reports whether Redo would move.
func (h *History[T]) CanRedo() bool { return len(h.future) > 0 }

// Change is one transition of a [StateMachine].
type Change[S, E any] struct {
	From, To S
	Event    E
}

// StateMachine is State: a finite state machine from a transition table.
//
//	order := gof.NewStateMachine("draft", map[string]map[string]string{
//		"draft": {"pay": "paid"},
//		"paid":  {"ship": "sent"},
//		"sent":  {},
//	})
//	order.Send("pay")   // "paid", nil
//	order.Send("pay")   // error: "pay" is not allowed in "sent"
//
// States and events are comparable type parameters, so a named string type or an
// int enum works and reads better than bare strings.
type StateMachine[S, E comparable] struct {
	state       S
	transitions map[S]map[E]S

	// Changes emits every transition, which is an audit trail without touching
	// the call sites.
	Changes Subject[Change[S, E]]
}

// NewStateMachine returns a machine in the initial state.
func NewStateMachine[S, E comparable](initial S, transitions map[S]map[E]S) *StateMachine[S, E] {
	return &StateMachine[S, E]{state: initial, transitions: transitions}
}

// State returns the current state.
func (m *StateMachine[S, E]) State() S { return m.state }

// Can reports whether event is allowed in the current state.
func (m *StateMachine[S, E]) Can(event E) bool {
	_, ok := m.transitions[m.state][event]
	return ok
}

// Send applies event, returning the new state, or an error naming the event and
// the state that refused it.
func (m *StateMachine[S, E]) Send(event E) (S, error) {
	target, ok := m.transitions[m.state][event]
	if !ok {
		return m.state, fmt.Errorf("gof: %v is not allowed in %v", event, m.state)
	}
	from := m.state
	m.state = target
	m.Changes.Emit(Change[S, E]{From: from, To: target, Event: event})
	return target, nil
}

// MustSend is Send for the case where an illegal transition is a programming error.
func (m *StateMachine[S, E]) MustSend(event E) S {
	state, err := m.Send(event)
	if err != nil {
		panic(err)
	}
	return state
}

// Visitor dispatches on a tag you extract from the node, rather than a switch
// repeated in every function that walks the tree.
//
// When your nodes are distinct Go types, a type switch is clearer and the compiler
// helps — use that. This is for a tag field, which is what a decoded JSON payload
// or a database row gives you.
func Visitor[Node any, Tag comparable, Result any](
	tag func(Node) Tag,
	visitors map[Tag]func(Node) Result,
	fallback func(Node) Result,
) func(Node) (Result, error) {
	return func(node Node) (Result, error) {
		key := tag(node)
		if visit, ok := visitors[key]; ok {
			return visit(node), nil
		}
		if fallback != nil {
			return fallback(node), nil
		}
		var zero Result
		return zero, fmt.Errorf("gof: no visitor for %v", key)
	}
}

// Iterate turns an external cursor into an [iter.Seq], so `for range` walks it.
//
// Drivers and SDKs hand you a HasNext/Next pair. This stays lazy: next runs only
// when the loop asks, so a `break` stops pulling. Writing the source yourself? A
// plain iter.Seq or a channel is simpler — this is for sources you did not write.
func Iterate[T any](hasNext func() bool, next func() T) iter.Seq[T] {
	return func(yield func(T) bool) {
		for hasNext() {
			if !yield(next()) {
				return
			}
		}
	}
}

// Template Method has no helper in Go. A struct of func fields with a constructor
// that fills the defaults is the whole pattern, and it is shorter than any helper:
//
//	type Report struct {
//		Read  func() string
//		Parse func(string) []string
//	}
//
//	func NewReport(options ...func(*Report)) *Report {
//		r := &Report{Read: readDB, Parse: parseCSV}
//		for _, option := range options { option(r) }
//		return r
//	}
//
// Functional options are the same idea applied to any constructor — [NewHistory]
// uses them. See the README table.

File: packages/ts/src/behavioral.ts
/** Behavioral patterns. */

/**
 * Chain of Responsibility — each handler either answers or calls `next()`.
 * Works for sync and async (`Res` can be a Promise).
 *
 * ```ts
 * const support = chain<Ticket, string>(
 *   [(t, next) => (t.level === 1 ? 'bot' : next()), (t, next) => (t.paid ? 'human' : next())],
 *   () => 'queue',
 * );
 * ```
 */
export type Handler<Req, Res> = (request: Req, next: () => Res) => Res;

export function chain<Req, Res>(
  handlers: Handler<Req, Res>[],
  fallback?: (request: Req) => Res,
): (request: Req) => Res {
  return (request) => {
    const step = (i: number): Res => {
      const handler = handlers[i];
      if (!handler) {
        if (!fallback) throw new Error('chain: no handler answered and no fallback was given');
        return fallback(request);
      }
      return handler(request, () => step(i + 1));
    };
    return step(0);
  };
}

/**
 * Command — undoable operations with history.
 *
 * ```ts
 * const bus = commandBus();
 * bus.run({ do: () => doc.push('a'), undo: () => doc.pop() });
 * bus.undo();
 * bus.redo();
 * ```
 * A command without `undo` clears the redo stack and cannot be undone.
 */
export type Command<T = unknown> = { do(): T; undo?(): void };

export type CommandBus = {
  run<T>(command: Command<T>): T;
  undo(): boolean;
  redo(): boolean;
  canUndo(): boolean;
  canRedo(): boolean;
};

export function commandBus(): CommandBus {
  const done: Command[] = [];
  let undone: Command[] = [];
  return {
    run(command) {
      const result = command.do();
      undone = [];
      if (command.undo) done.push(command as Command);
      return result;
    },
    undo() {
      const command = done.pop();
      if (!command) return false;
      command.undo?.();
      undone.push(command);
      return true;
    },
    redo() {
      const command = undone.pop();
      if (!command) return false;
      command.do();
      done.push(command);
      return true;
    },
    canUndo: () => done.length > 0,
    canRedo: () => undone.length > 0,
  };
}

/**
 * Observer — one typed channel. `subscribe` returns the unsubscribe function.
 *
 * ```ts
 * const priceChanged = subject<number>();
 * const off = priceChanged.subscribe((p) => render(p));
 * priceChanged.emit(9.99);
 * off();
 * ```
 */
export type Subject<T> = {
  subscribe(listener: (value: T) => void): () => void;
  emit(value: T): void;
  size(): number;
};

export function subject<T>(): Subject<T> {
  const listeners = new Set<(value: T) => void>();
  return {
    subscribe(listener) {
      listeners.add(listener);
      return () => void listeners.delete(listener);
    },
    // Iterate a copy: a listener may unsubscribe during emit.
    emit: (value) => [...listeners].forEach((listener) => listener(value)),
    size: () => listeners.size,
  };
}

/**
 * Mediator — components talk to a typed hub, never to each other.
 *
 * ```ts
 * const hub = mediator<{ login: { id: string }; logout: void }>();
 * hub.on('login', ({ id }) => track(id));
 * hub.emit('login', { id: 'u1' });
 * ```
 */
export type Mediator<E extends Record<string, unknown>> = {
  on<K extends keyof E>(event: K, listener: (payload: E[K]) => void): () => void;
  emit<K extends keyof E>(event: K, payload: E[K]): void;
};

export function mediator<E extends Record<string, unknown>>(): Mediator<E> {
  const channels = new Map<keyof E, Subject<E[keyof E]>>();
  const channel = <K extends keyof E>(event: K) => {
    let existing = channels.get(event);
    if (!existing) channels.set(event, (existing = subject()));
    return existing as Subject<E[K]>;
  };
  return {
    on: (event, listener) => channel(event).subscribe(listener),
    emit: (event, payload) => channel(event).emit(payload),
  };
}

/**
 * Memento — undo/redo over snapshots of state.
 *
 * ```ts
 * const h = history({ text: '' });
 * h.save({ text: 'hi' });
 * h.undo(); // { text: '' }
 * ```
 * Snapshots are stored by reference — pass a copy, or a `snapshot` function,
 * if the state object is mutated in place.
 */
export type History<T> = {
  current(): T;
  save(state: T): void;
  undo(): T | undefined;
  redo(): T | undefined;
  canUndo(): boolean;
  canRedo(): boolean;
};

export function history<T>(
  initial: T,
  // `NoInfer` on snapshot matters: without it, passing a generic function such as
  // `structuredClone` makes T infer as `{}` and the history loses its type. Found
  // by writing examples/07-undo, where it turned `.undo()?.tags` into an error.
  options: { limit?: number; snapshot?: (state: NoInfer<T>) => NoInfer<T> } = {},
): History<T> {
  const { limit = Infinity, snapshot = (state: T) => state } = options;
  const past: T[] = [];
  const future: T[] = [];
  let present = snapshot(initial);
  return {
    current: () => present,
    save(state) {
      past.push(present);
      if (past.length > limit) past.shift();
      future.length = 0;
      present = snapshot(state);
    },
    undo() {
      if (!past.length) return undefined;
      future.push(present);
      return (present = past.pop() as T);
    },
    redo() {
      if (!future.length) return undefined;
      past.push(present);
      return (present = future.pop() as T);
    },
    canUndo: () => past.length > 0,
    canRedo: () => future.length > 0,
  };
}

/**
 * State — a finite state machine from a transition table.
 *
 * ```ts
 * const order = stateMachine({
 *   initial: 'draft',
 *   states: { draft: { pay: 'paid' }, paid: { ship: 'sent' }, sent: {} },
 * });
 * order.send('pay');  // 'paid'
 * order.can('ship');  // true
 * ```
 * A transition value may be a target state or `(payload) => target`.
 */
export type Transitions<S extends string, E extends string> = Record<
  S,
  Partial<Record<E, S | ((payload: unknown) => S)>>
>;

export type StateMachine<S extends string, E extends string> = {
  state(): S;
  can(event: E): boolean;
  /** Returns the new state. Throws on an event the current state does not allow. */
  send(event: E, payload?: unknown): S;
  onChange(listener: (change: { from: S; to: S; event: E }) => void): () => void;
};

export function stateMachine<S extends string, E extends string>(config: {
  initial: S;
  states: Transitions<S, E>;
}): StateMachine<S, E> {
  let state = config.initial;
  const changes = subject<{ from: S; to: S; event: E }>();
  const target = (event: E) => config.states[state]?.[event];
  return {
    state: () => state,
    can: (event) => target(event) !== undefined,
    send(event, payload) {
      const to = target(event);
      if (to === undefined) throw new Error(`stateMachine: "${event}" is not allowed in "${state}"`);
      const from = state;
      // S extends string, so a function value is always the dynamic-target form.
      state = typeof to === 'function' ? (to as (payload: unknown) => S)(payload) : (to as S);
      changes.emit({ from, to: state, event });
      return state;
    },
    onChange: (listener) => changes.subscribe(listener),
  };
}

/**
 * Visitor — dispatch on a node's kind instead of a switch scattered everywhere.
 *
 * ```ts
 * const area = visitor<Shape, number>({
 *   circle: (c) => Math.PI * c.r ** 2,
 *   square: (s) => s.side ** 2,
 * });
 * area({ type: 'circle', r: 1 });
 * ```
 */
export function visitor<N extends { type: string }, R>(
  visitors: { [K in N['type']]?: (node: Extract<N, { type: K }>) => R },
  fallback?: (node: N) => R,
): (node: N) => R {
  return (node) => {
    const visit = visitors[node.type as N['type']] as ((n: N) => R) | undefined;
    if (visit) return visit(node);
    if (fallback) return fallback(node);
    throw new Error(`visitor: no visitor for "${node.type}"`);
  };
}

/**
 * Template Method — a fixed skeleton with replaceable steps.
 *
 * ```ts
 * const mine = template({ read: () => csv(), parse: (s: string) => s.split(',') },
 *   (hooks) => hooks.parse(hooks.read()));
 * mine({ read: () => 'a,b' })(); // ['a', 'b']
 * ```
 */
export function template<H extends Record<string, (...args: never[]) => unknown>, A extends unknown[], R>(
  defaults: H,
  skeleton: (hooks: H, ...args: A) => R,
): (overrides?: Partial<H>) => (...args: A) => R {
  return (overrides) => (...args) => skeleton({ ...defaults, ...overrides }, ...args);
}

// ponytail: Iterator is a generator function plus `for...of` — already in the
// language, so there is no helper for it.

/**
 * Iterator — turn an external cursor into something `for...of` can walk.
 *
 * Database cursors, paginated HTTP APIs and vendor SDKs hand you a
 * `hasNext()` / `next()` pair. This makes that a real iterable, so spread,
 * destructuring, `for...of` and early `break` all work:
 *
 * ```ts
 * for (const row of iterate(dbCursor)) {
 *   if (row.id === target) break;   // the cursor stops here
 * }
 * ```
 *
 * Lazy: `next()` is called only when the consumer asks for the next value. If you
 * are writing the source yourself, a generator function is simpler — this is for
 * sources you did not write.
 */
export function iterate<T>(source: { hasNext(): boolean; next(): T }): IterableIterator<T> {
  function* walk(): Generator<T> {
    while (source.hasNext()) yield source.next();
  }
  return walk();
}
